lib/internal/source_map/source_map_cache_map.js

```diff
@@ -0,0 +1,110 @@
+'use strict';
+
+const {
+  ArrayPrototypeForEach,
+  ObjectFreeze,
+  SafeFinalizationRegistry,
+  SafeMap,
+  SafeWeakMap,
+  SafeWeakRef,
+  SymbolIterator,
+} = primordials;
+
+/**
+ * Specialized WeakMap that caches source map entries by `filename` and `sourceURL`.
+ * Cached entries can be iterated with `for..of`.
+ *
+ * The cache map maintains the cache entries by:
+ * - `weakTargetMap`(Map): a strong sourceURL -> WeakRef(Module),
+ * - `weakMap`(WeakMap): a Module instance object -> source map data.
+ *
+ * Obsolete `weakTargetMap` entries are removed by the `finalizationRegistry` callback.
+ * This pattern decouples the strong url reference to the source map data and allow the
+ * cache to be reclaimed eagerly, without depending on a undeterministic callback of a
+ * finalization registry.
+ */
+class SourceMapCacheMap {
+  /**
+   * @type {Map<string, WeakRef<*>>}
+   * The cached module instance can be removed from the global module registry
+   * with approaches like mutating `require.cache`.
+   * The `weakTargetMap` exposes entries by `filename` and `sourceURL`.
+   * In the case of mutated module registry, obsolete entries are removed from
+   * the cache by the `finalizationRegistry`.
+   */
+  #weakTargetMap = new SafeMap();
+  #weakMap = new SafeWeakMap();
```

Comment:
Actually, I think we can just take advantage of the weakmap semantics and use a private symbol to reference the metadata from the module instances. That's typically more GC-efficient than the ephemeron GC required by the weak map.

Reply:
Thanks, it makes sense to replace the WeakMap with a private symbol instead. Updated!